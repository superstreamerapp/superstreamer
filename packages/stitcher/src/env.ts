import { getEnv } from "shared/env";
import { z } from "zod";

export const env = getEnv(
  z.object({
    PORT: z.coerce.number().default(52002),

    KV: z.enum(["memory", "redis", "cloudflare-kv"]).default("redis"),
    REDIS_HOST: z.string().default("localhost"),
    REDIS_PORT: z.coerce.number().default(6379),

    PUBLIC_S3_ENDPOINT: z.string(),
    PUBLIC_STITCHER_ENDPOINT: z.string(),
    PUBLIC_API_ENDPOINT: z.string().optional(),

    // Secret is optional, if we don't provide it, we won't be able to
    // call API requests but one might not need to do that.
    SUPER_SECRET: z.string().optional(),
  }),
);
