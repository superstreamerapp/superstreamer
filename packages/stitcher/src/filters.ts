import { t } from "elysia";
import type { MasterPlaylist } from "./parser";

export interface Filter {
  resolution?: string;
  audioLanguage?: string;
}

export function formatFilterToQueryParam(filter?: Filter) {
  if (!filter) {
    filter = {};
  }
  return btoa(JSON.stringify(filter));
}

export const filterSchema = t.Optional(
  t
    .Transform(t.String())
    .Decode((value) => JSON.parse(atob(value)) as Filter)
    .Encode((filter) => btoa(JSON.stringify(filter))),
);

function parseRange(input: string): [number, number] | null {
  const match = input.match(/^(\d+)-(\d+)$/);

  if (match?.[1] && match[2]) {
    const min = parseInt(match[1]);
    const max = parseInt(match[2]);
    return [min, max];
  }

  return null;
}

function parseOperatorToRange(input: string): [number, number] | null {
  const match = input.match(/(<=?|>=?)\s*(\d+)/);
  if (match?.[2] === undefined) {
    return null;
  }

  const operator = match[1];
  const number = parseInt(match[2]);

  if (operator === "<=") {
    return [0, number];
  } else if (operator === "<") {
    return [0, number - 1];
  } else if (operator === ">=") {
    return [number, Infinity];
  } else if (operator === ">") {
    return [number + 1, Infinity];
  }

  return null;
}

function parseFilterToRange(input: string): [number, number] {
  let range = parseRange(input);
  if (range) {
    return range;
  }

  range = parseOperatorToRange(input);
  if (range) {
    return range;
  }

  throw new Error(`Failed to parse to range "${input}"`);
}

function parseFilterToList(input: string) {
  return input.split(",").map((value) => value.trim());
}

export function filterMasterPlaylist(master: MasterPlaylist, filter: Filter) {
  if (filter.resolution !== undefined) {
    const [min, max] = parseFilterToRange(filter.resolution);
    master.variants = master.variants.filter(
      (variant) =>
        // If we have no height, we'll make it pass.
        !variant.resolution?.height ||
        // If the variant height is within our range.
        (variant.resolution.height >= min && variant.resolution.height <= max),
    );
  }
  if (filter.audioLanguage !== undefined) {
    const list = parseFilterToList(filter.audioLanguage);
    master.renditions = master.renditions.filter((rendition) => {
      if (rendition.type === "AUDIO") {
        if (rendition.language && list.includes(rendition.language)) {
          return true;
        }
        return false;
      }
      return true;
    });
  }
}
