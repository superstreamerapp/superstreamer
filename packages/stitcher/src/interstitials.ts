import { DateTime } from "luxon";
import { HashGroup } from "./lib/hash-group";
import { createUrl, replaceUrlParams } from "./lib/url";
import { getAssetsFromVastData, getAssetsFromVastUrl } from "./vast";
import type { DateRange, Segment } from "./parser";
import type { Session } from "./session";
import type { Asset } from "./types";

// An item describes what we'd like to collect for a particular date.
interface DateGroupItem {
  timelineStyle: "HIGHLIGHT" | "PRIMARY";
  replaceContent: boolean;
  maxDuration?: number;
}

export function getStaticDateRanges(
  session: Session,
  segments: Segment[],
  isLive: boolean,
) {
  const dateGroup = new HashGroup<number, DateGroupItem>({
    getDefaultValue: () => ({
      timelineStyle: "PRIMARY",
      replaceContent: isLive ? true : false,
    }),
  });

  // Collect dateRanges from splice points.
  for (const segment of segments) {
    if (segment.spliceInfo?.type !== "OUT" || !segment.programDateTime) {
      continue;
    }

    const key = segment.programDateTime.toMillis();
    const item = dateGroup.get(key);

    item.timelineStyle = "HIGHLIGHT";
    item.maxDuration = segment.spliceInfo.duration;
  }

  // Collect dateRanges from each event defined in the session.
  for (const event of session.events) {
    const key = event.dateTime.toMillis();
    const item = dateGroup.get(key);

    if (event.vast) {
      // If we resolved the event by a vast, we know it's an ad and can mark it
      // as HIGHLIGHT on the timeline.
      item.timelineStyle = "HIGHLIGHT";
    }

    if (
      !event.maxDuration ||
      (item.maxDuration && event.maxDuration > item.maxDuration)
    ) {
      // If we have a max duration for this event, we'll save it for this interstitial. Always takes the
      // largest maxDuration across events.
      item.maxDuration = event.maxDuration;
    }
  }

  return dateGroup.toEntries().map<DateRange>(([key, item]) => {
    const startDate = DateTime.fromMillis(key);

    const assetListUrl = createUrl("out/asset-list.json", {
      dt: startDate.toISO(),
      sid: session.id,
      mdur: item.maxDuration,
    });

    const clientAttributes: Record<string, number | string> = {
      RESTRICT: "SKIP,JUMP",
      "ASSET-LIST": assetListUrl,
      "CONTENT-MAY-VARY": "YES",
      "TIMELINE-OCCUPIES": "POINT",
      "TIMELINE-STYLE": item.timelineStyle,
    };

    if (!item.replaceContent) {
      clientAttributes["RESUME-OFFSET"] = 0;
    }

    if (item.maxDuration !== undefined) {
      clientAttributes["PLAYOUT-LIMIT"] = item.maxDuration;
    }

    const cue: string[] = [];
    if (startDate.equals(session.startTime)) {
      cue.push("PRE");
    }

    if (cue.length) {
      clientAttributes["CUE"] = cue.join(",");
    }

    return {
      classId: "com.apple.hls.interstitial",
      id: `${startDate.toMillis()}`,
      startDate,
      clientAttributes,
    };
  });
}

export async function getAssets(
  session: Session,
  dateTime: DateTime,
  maxDuration?: number,
): Promise<Asset[]> {
  // Filter all events for a particular dateTime, we'll need to transform these to
  // a list of assets.
  const events = session.events.filter((event) =>
    event.dateTime.equals(dateTime),
  );

  const assets: Asset[] = [];

  for (const event of events) {
    if (event.vast) {
      const { url, data } = event.vast;

      // The event contains a VAST url.
      if (url) {
        const vastUrl = replaceUrlParams(url, {
          maxDuration,
        });
        const vastAssets = await getAssetsFromVastUrl(vastUrl);
        assets.push(...vastAssets);
      }

      // The event contains inline VAST data.
      if (data) {
        const vastAssets = await getAssetsFromVastData(data);
        assets.push(...vastAssets);
      }
    }

    // The event contains a list of assets, explicitly defined.
    if (event.assets) {
      assets.push(...event.assets);
    }
  }

  // If we have a generic vast config on our session, use that one to resolve (eg; for live streams)
  if (session.vast?.url) {
    const vastUrl = replaceUrlParams(session.vast.url, {
      maxDuration,
    });
    const tempAssets = await getAssetsFromVastUrl(vastUrl);
    assets.push(...tempAssets);
  }

  return assets;
}
