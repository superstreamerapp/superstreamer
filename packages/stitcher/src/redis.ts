import { env } from "hono/adapter";
import { createClient } from "redis";
import type { Context } from "hono";
import type { RedisClientType } from "redis";

let redis: RedisClientType | null = null;

const REDIS_PREFIX = "stitcher";

async function getRedis(context: Context) {
  if (!redis) {
    const { REDIS_HOST, REDIS_PORT } = env<{
      REDIS_HOST: string;
      REDIS_PORT: string;
    }>(context);

    redis = createClient({
      socket: {
        host: REDIS_HOST,
        port: +REDIS_PORT,
      },
    });

    await redis.connect();
  }
  return redis;
}

export async function getRedisKey(context: Context, key: string) {
  const redis = await getRedis(context);
  return await redis.get(`${REDIS_PREFIX}:${key}`);
}

export async function setRedisKeyValue(
  context: Context,
  key: string,
  value: string,
  ttl: number,
) {
  const redis = await getRedis(context);
  await redis.set(`${REDIS_PREFIX}:${key}`, value, {
    EX: ttl,
  });
}
