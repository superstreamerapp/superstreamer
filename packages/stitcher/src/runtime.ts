export interface AppVariables {
  env: {
    REDIS_HOST?: string;
    REDIS_PORT?: string;
    PUBLIC_S3_ENDPOINT: string;
    PUBLIC_STITCHER_ENDPOINT: string;
    PUBLIC_API_ENDPOINT?: string;
    SUPER_SECRET?: string;
  };
}
