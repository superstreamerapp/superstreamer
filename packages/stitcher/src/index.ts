import { Hono } from "hono";
import { cors } from "hono/cors";
import { openAPISpecs } from "hono-openapi";
import { env } from "./env";
import { outApp } from "./routes/out";
import { sessionApp } from "./routes/session";

const app = new Hono()
  .use(cors())
  .route("/session", sessionApp)
  .route("/out", outApp);

app.get(
  "/openapi",
  openAPISpecs(app, {
    documentation: {
      info: {
        title: "Superstreamer Stitcher API",
        version: "1.0.0",
        description:
          "Realtime playlist manipulator. Can be used for ad, bumper or other HLS interstitials insertion on-the-fly. Can apply filters to playlists.",
      },
    },
  }),
);

export default {
  ...app,
  port: env.PORT,
};
