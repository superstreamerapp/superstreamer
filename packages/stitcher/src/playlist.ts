import { assert } from "shared/assert";
import { filterMasterPlaylist, formatFilterToQueryParam } from "./filters";
import { getAssets, getStaticDateRanges } from "./interstitials";
import { encrypt } from "./lib/crypto";
import { createUrl, joinUrl, replaceUrlParams } from "./lib/url";
import {
  parseMasterPlaylist,
  parseMediaPlaylist,
  stringifyMasterPlaylist,
  stringifyMediaPlaylist,
} from "./parser";
import { updateSession } from "./session";
import { fetchVmap, toAdBreakTimeOffset } from "./vmap";
import type { Filter } from "./filters";
import type { MasterPlaylist, MediaPlaylist } from "./parser";
import type { Session } from "./session";
import type { TimedEvent } from "./types";
import type { VmapAdBreak } from "./vmap";
import type { DateTime } from "luxon";

export async function formatMasterPlaylist(params: {
  origUrl: string;
  session: Session;
  filter?: Filter;
}) {
  await initSessionOnMasterReq(params.session);

  const master = await fetchMasterPlaylist(params.origUrl);

  if (params.filter) {
    filterMasterPlaylist(master, params.filter);
  }

  rewriteMasterPlaylistUrls(master, params);

  return stringifyMasterPlaylist(master);
}

export async function formatMediaPlaylist(
  mediaUrl: string,
  session: Session,
  type: "video" | "audio" | "subtitles",
) {
  const media = await fetchMediaPlaylist(mediaUrl);

  const firstSegment = media.segments[0];
  assert(firstSegment);

  if (media.endlist) {
    firstSegment.programDateTime = session.startTime;
  }

  // Apply dateRanges to each video playlist.
  if (type === "video") {
    // If we have an endlist and a PDT, we can add static date ranges based on this.
    const isLive = !media.endlist;

    media.dateRanges = getStaticDateRanges(session, media.segments, isLive);
  }

  rewriteSpliceInfoSegments(media);

  rewriteMediaPlaylistUrls(media, mediaUrl);

  return stringifyMediaPlaylist(media);
}

export async function formatAssetList(
  session: Session,
  dateTime: DateTime,
  maxDuration?: number,
) {
  const assets = await getAssets(session, dateTime, maxDuration);

  return {
    ASSETS: assets.map((asset) => {
      return {
        URI: asset.url,
        DURATION: asset.duration,
      };
    }),
  };
}

async function fetchMasterPlaylist(url: string) {
  const response = await fetch(url);
  const result = await response.text();
  return parseMasterPlaylist(result);
}

async function fetchMediaPlaylist(url: string) {
  const response = await fetch(url);
  const result = await response.text();
  return parseMediaPlaylist(result);
}

export async function fetchDuration(url: string) {
  const variant = (await fetchMasterPlaylist(url))?.variants[0];

  if (!variant) {
    throw new Error(`Missing variant for "${url}"`);
  }

  const media = await fetchMediaPlaylist(joinUrl(url, variant.uri));
  return media.segments.reduce((acc, segment) => {
    acc += segment.duration;
    return acc;
  }, 0);
}

export function createMasterUrl(params: {
  url: string;
  session: Session;
  filter?: Filter;
}) {
  const fil = formatFilterToQueryParam(params.filter);

  const url = createUrl("out/master.m3u8", {
    eurl: encrypt(params.url),
    sid: params.session.id,
    fil,
  });

  return url;
}

function createMediaUrl(params: {
  url: string;
  sessionId: string;
  type: "video" | "audio" | "subtitles";
}) {
  return createUrl("out/playlist.m3u8", {
    eurl: encrypt(params.url),
    sid: params.sessionId,
    type: params.type,
  });
}

export function rewriteMasterPlaylistUrls(
  master: MasterPlaylist,
  params: {
    origUrl: string;
    session: Session;
  },
) {
  for (const variant of master.variants) {
    const url = joinUrl(params.origUrl, variant.uri);
    variant.uri = createMediaUrl({
      url,
      sessionId: params.session.id,
      type: "video",
    });
  }

  for (const rendition of master.renditions) {
    if (!rendition.uri) {
      continue;
    }

    const url = joinUrl(params.origUrl, rendition.uri);

    let type: "audio" | "subtitles" | undefined;
    if (rendition.type === "AUDIO") {
      type = "audio";
    } else if (rendition.type === "SUBTITLES") {
      type = "subtitles";
    }

    if (!type) {
      continue;
    }

    rendition.uri = createMediaUrl({
      url,
      sessionId: params.session.id,
      type,
    });
  }
}

export function rewriteMediaPlaylistUrls(
  media: MediaPlaylist,
  mediaUrl: string,
) {
  media.segments.forEach((segment) => {
    if (segment.map?.uri === "init.mp4") {
      segment.map.uri = joinUrl(mediaUrl, segment.map.uri);
    }
    segment.uri = joinUrl(mediaUrl, segment.uri);
  });
}

/**
 * Go over each segment in a media playlist and swap the splice info for a discontinuity.
 * @param media
 */
export function rewriteSpliceInfoSegments(media: MediaPlaylist) {
  media.segments.forEach((segment) => {
    if (segment.spliceInfo) {
      delete segment.spliceInfo;
      segment.discontinuity = true;
    }
  });
}

async function initSessionOnMasterReq(session: Session) {
  let storeSession = false;

  // If we have a vmap config but no result yet, we'll resolve it.
  if (session.vmap && !session.vmap.result) {
    const vmapUrl = replaceUrlParams(session.vmap.url);
    const vmap = await fetchVmap(vmapUrl);

    // Store an empty object, if we want vmap specific info to be stored
    // across sessions later, we can do it here, such as tracking pixels.
    session.vmap.result = {};

    vmap.adBreaks.forEach((adBreak) => {
      const event = mapAdBreakToTimedEvent(session.startTime, adBreak);
      if (event) {
        session.events.push(event);
      }
    });

    storeSession = true;
  }

  if (storeSession) {
    await updateSession(session);
  }
}

export function mapAdBreakToTimedEvent(
  startTime: DateTime,
  adBreak: VmapAdBreak,
): TimedEvent | null {
  const timeOffset = toAdBreakTimeOffset(adBreak);

  if (timeOffset === null) {
    return null;
  }

  const dateTime = startTime.plus({ seconds: timeOffset });

  return {
    dateTime,
    vast: {
      url: adBreak.vastUrl,
      data: adBreak.vastData,
    },
  };
}
