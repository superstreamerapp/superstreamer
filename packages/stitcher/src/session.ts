import { randomUUID } from "crypto";
import { DateTime } from "luxon";
import { kv } from "./adapters/kv";
import { mergeInterstitials } from "./interstitials";
import { JSON } from "./lib/json";
import { resolveUri } from "./lib/url";
import { fetchDuration } from "./playlist";
import type { Interstitial, InterstitialChunk } from "./types";
import type { VmapParams } from "./vmap";

export interface Session {
  id: string;
  url: string;
  expiry: number;
  startTime: DateTime;

  // User defined options
  vmap?: VmapParams;
  interstitials: Interstitial[];
}

interface SessionInterstitial {
  time: number | string;
  duration?: number;
  assets?: {
    uri: string;
    kind?: "ad" | "bumper";
  }[];
  vast?: {
    url: string;
  };
  assetList?: {
    url: string;
  };
}

export async function createSession(params: {
  uri: string;
  vmap?: VmapParams;
  interstitials?: SessionInterstitial[];
  expiry?: number;
}) {
  const id = randomUUID();
  const startTime = DateTime.now();

  const session: Session = {
    id,
    url: resolveUri(params.uri),
    vmap: params.vmap,
    startTime,
    interstitials: [],
    // A session is valid for 3 hours by default.
    expiry: params.expiry ?? 60 * 60 * 3,
  };

  if (params.interstitials) {
    const interstitials = await mapSessionInterstitials(
      startTime,
      params.interstitials,
    );
    mergeInterstitials(session.interstitials, interstitials);
  }

  const value = JSON.stringify(session);
  await kv.set(`session:${id}`, value, session.expiry);

  return session;
}

export async function getSession(id: string) {
  const data = await kv.get(`session:${id}`);
  if (!data) {
    throw new Error(`No session found for id ${id}`);
  }
  return JSON.parse<Session>(data);
}

export async function updateSession(session: Session) {
  const value = JSON.stringify(session);
  await kv.set(`session:${session.id}`, value, session.expiry);
}

async function mapSessionInterstitials(
  startTime: DateTime,
  interstitials: SessionInterstitial[],
) {
  const result: Interstitial[] = [];

  for (const interstitial of interstitials) {
    const dateTime =
      typeof interstitial.time === "string"
        ? DateTime.fromISO(interstitial.time)
        : startTime.plus({ seconds: interstitial.time });

    const chunks: InterstitialChunk[] = [];

    if (interstitial.assets) {
      for (const asset of interstitial.assets) {
        const { uri, kind } = asset;

        const url = resolveUri(uri);

        chunks.push({
          type: "asset",
          data: {
            url,
            duration: await fetchDuration(url),
            kind,
          },
        });
      }
    }

    if (interstitial.vast) {
      chunks.push({ type: "vast", data: interstitial.vast });
    }

    if (interstitial.assetList) {
      chunks.push({ type: "assetList", data: interstitial.assetList });
    }

    result.push({
      dateTime,
      duration: interstitial.duration,
      chunks,
    });
  }

  return result;
}
