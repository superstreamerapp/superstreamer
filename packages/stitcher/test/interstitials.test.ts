import { describe, expect, spyOn, test } from "bun:test";
import { DateTime } from "luxon";
import { addFakeInterstitials, fakeSession } from "./test-data";
import { getAssets, getStaticDateRanges } from "../src/interstitials";

describe("getStaticDateRanges", () => {
  test("should create dateRanges for vod", () => {
    const session = fakeSession();
    addFakeInterstitials(session);

    const isLive = false;
    const dateRanges = getStaticDateRanges(session, isLive);

    expect(dateRanges).toMatchSnapshot();
  });

  test("should create dateRanges for live", () => {
    const session = fakeSession();
    addFakeInterstitials(session);

    const isLive = true;
    const dateRanges = getStaticDateRanges(session, isLive);

    expect(dateRanges).toMatchSnapshot();
  });
});

describe("getAssets", () => {
  test("should get assets by interstitials", async () => {
    const session = fakeSession();
    const dateTime = DateTime.now();

    const spy = spyOn(await import("../src/vast"), "getAssetsFromVast");
    spy.mockReturnValueOnce(
      Promise.resolve([
        {
          url: "https://mock.com/ad_1/master.m3u8",
          duration: 25,
        },
      ]),
    );

    session.interstitials = [
      {
        dateTime,
        chunks: [
          {
            type: "asset",
            data: {
              url: "https://mock.com/interstitial1/master.m3u8",
              duration: 10,
              kind: "ad",
            },
          },
          {
            type: "vast",
            data: {
              url: "https://mock.com/vast.xml",
            },
          },
        ],
      },
    ];

    const assets = await getAssets(session, dateTime);

    expect(spy).toHaveBeenCalledTimes(1);
    expect(assets).toMatchSnapshot();

    spy.mockRestore();
  });
});
