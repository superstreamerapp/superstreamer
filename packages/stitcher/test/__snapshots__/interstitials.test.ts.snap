// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`getStaticDateRanges should create dateRanges for vod 1`] = `
[
  {
    "classId": "com.apple.hls.interstitial",
    "clientAttributes": {
      "ASSET-LIST": "stitcher-endpoint/out/asset-list.json?dt=2021-05-02T10%3A12%3A05.250%2B00%3A00&sid=36bab417-0952-4c23-bdf0-9a424e4651ad",
      "CONTENT-MAY-VARY": "YES",
      "CUE": "ONCE,PRE",
      "RESTRICT": "SKIP,JUMP",
      "RESUME-OFFSET": 0,
      "TIMELINE-OCCUPIES": "POINT",
      "TIMELINE-STYLE": "HIGHLIGHT",
    },
    "id": "1619950325250",
    "startDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 10,
        "millisecond": 250,
        "minute": 12,
        "month": 5,
        "second": 5,
        "year": 2021,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1619950325250,
      "weekData": null,
    },
  },
  {
    "classId": "com.apple.hls.interstitial",
    "clientAttributes": {
      "ASSET-LIST": "stitcher-endpoint/out/asset-list.json?dt=2021-05-02T10%3A12%3A15.250%2B00%3A00&sid=36bab417-0952-4c23-bdf0-9a424e4651ad",
      "CONTENT-MAY-VARY": "YES",
      "CUE": "ONCE",
      "RESTRICT": "SKIP,JUMP",
      "RESUME-OFFSET": 0,
      "TIMELINE-OCCUPIES": "POINT",
      "TIMELINE-STYLE": "HIGHLIGHT",
    },
    "id": "1619950335250",
    "startDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 10,
        "millisecond": 250,
        "minute": 12,
        "month": 5,
        "second": 15,
        "year": 2021,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1619950335250,
      "weekData": null,
    },
  },
  {
    "classId": "com.apple.hls.interstitial",
    "clientAttributes": {
      "ASSET-LIST": "stitcher-endpoint/out/asset-list.json?dt=2021-05-02T10%3A12%3A35.250%2B00%3A00&sid=36bab417-0952-4c23-bdf0-9a424e4651ad",
      "CONTENT-MAY-VARY": "YES",
      "CUE": "ONCE",
      "RESTRICT": "SKIP,JUMP",
      "RESUME-OFFSET": 0,
      "TIMELINE-OCCUPIES": "POINT",
      "TIMELINE-STYLE": "HIGHLIGHT",
    },
    "id": "1619950355250",
    "startDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 10,
        "millisecond": 250,
        "minute": 12,
        "month": 5,
        "second": 35,
        "year": 2021,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1619950355250,
      "weekData": null,
    },
  },
]
`;

exports[`getStaticDateRanges should create dateRanges for live 1`] = `
[
  {
    "classId": "com.apple.hls.interstitial",
    "clientAttributes": {
      "ASSET-LIST": "stitcher-endpoint/out/asset-list.json?dt=2021-05-02T10%3A12%3A05.250%2B00%3A00&sid=36bab417-0952-4c23-bdf0-9a424e4651ad",
      "CONTENT-MAY-VARY": "YES",
      "CUE": "ONCE,PRE",
      "RESTRICT": "SKIP,JUMP",
      "TIMELINE-OCCUPIES": "POINT",
      "TIMELINE-STYLE": "HIGHLIGHT",
    },
    "id": "1619950325250",
    "startDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 10,
        "millisecond": 250,
        "minute": 12,
        "month": 5,
        "second": 5,
        "year": 2021,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1619950325250,
      "weekData": null,
    },
  },
  {
    "classId": "com.apple.hls.interstitial",
    "clientAttributes": {
      "ASSET-LIST": "stitcher-endpoint/out/asset-list.json?dt=2021-05-02T10%3A12%3A15.250%2B00%3A00&sid=36bab417-0952-4c23-bdf0-9a424e4651ad",
      "CONTENT-MAY-VARY": "YES",
      "CUE": "ONCE",
      "RESTRICT": "SKIP,JUMP",
      "TIMELINE-OCCUPIES": "POINT",
      "TIMELINE-STYLE": "HIGHLIGHT",
    },
    "id": "1619950335250",
    "startDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 10,
        "millisecond": 250,
        "minute": 12,
        "month": 5,
        "second": 15,
        "year": 2021,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1619950335250,
      "weekData": null,
    },
  },
  {
    "classId": "com.apple.hls.interstitial",
    "clientAttributes": {
      "ASSET-LIST": "stitcher-endpoint/out/asset-list.json?dt=2021-05-02T10%3A12%3A35.250%2B00%3A00&sid=36bab417-0952-4c23-bdf0-9a424e4651ad",
      "CONTENT-MAY-VARY": "YES",
      "CUE": "ONCE",
      "RESTRICT": "SKIP,JUMP",
      "TIMELINE-OCCUPIES": "POINT",
      "TIMELINE-STYLE": "HIGHLIGHT",
    },
    "id": "1619950355250",
    "startDate": DateTime {
      "_zone": SystemZone {},
      "c": {
        "day": 2,
        "hour": 10,
        "millisecond": 250,
        "minute": 12,
        "month": 5,
        "second": 35,
        "year": 2021,
      },
      "invalid": null,
      "isLuxonDateTime": true,
      "loc": Locale {
        "eraCache": {},
        "fastNumbersCached": null,
        "intl": "en-US",
        "locale": "en-US",
        "meridiemCache": null,
        "monthsCache": {
          "format": {},
          "standalone": {},
        },
        "numberingSystem": null,
        "outputCalendar": null,
        "specifiedLocale": null,
        "weekSettings": null,
        "weekdaysCache": {
          "format": {},
          "standalone": {},
        },
      },
      "localWeekData": null,
      "o": -0,
      "ts": 1619950355250,
      "weekData": null,
    },
  },
]
`;

exports[`getAssets should get assets by interstitials 1`] = `
[
  {
    "duration": 10,
    "kind": "ad",
    "url": "https://mock.com/interstitial1/master.m3u8",
  },
  {
    "duration": 25,
    "url": "https://mock.com/ad_1/master.m3u8",
  },
]
`;
